en/dev_log/log_encryption.md

```diff
@@ -0,0 +1,223 @@
+
+The encrypted log file and encrypted symmetric key file are displayed with a timestamp (but no filename) in QGroundControl, as shown below.
+You can determine which files are associated based on their timestamps.
+
+![QGroundControl ULog Download](../../assets/qgc/analyze/encrypted_log.png)
+
+Select and download both files.
+
+Note that both files will be downloaded with the `.ulg` suffix.
+You can identify the symmetric key file, as it is usually much smaller than the log file (about 300 bytes)
+
+For convenience in the decryption step, you might rename the file extensions to add back the `.ulgc` (log) and `.ulgk` (key) file extensions.
+
+### Decrypt ULogs
+
+Before you can analyze your encrypted logs, you will need to decrypt them.
+There is a Python script that can be used to decrypt logs in `Tools/decrypt_ulog.py`.
+
+`decrypt_ulog.py` takes 3 arguments:
+
+1. The encrypted `.ulogc` file.
+2. The symmetric key `.ulogk` file.
+3. The decryption key (the RSA2048 `.pem` private key which is used to unwrap the `.ulogk` file).
+
+```sh
+usage: decrypt_ulog.py [-h] [ulog_file] [ulog_key] [rsa_key]
+
+CLI tool to decrypt an ulog file
+
+positional arguments:
+  ulog_file   .ulog file
+  ulog_key    .ulogk, encrypted key
+  rsa_key     .pem format key for decrypting the ulog key
+
+optional arguments:
+  -h, --help  show this help message and exit
+
+```
+
+As an example:
+
+```sh
+python3 decrypt_ulog.py \
+/home/john/Downloads/log_24_2024-10-6-23-39-50.ulgc \
+/home/john/Downloads/log_23_2024-10-6-23-39-48.ulgk \
+new_keys/private_key.pem
+```
+
+On success the decrypted log file is created with the `.ul` suffix instead of `.ulg`.
+Rename the file back to `.ulg` and it is now ready for flight review.
+
+## Generate RSA Public & Private Keys
+
+In a production environment, you should your own generated keys rather than using the example keys.
+
+To generate the `CONFIG_PUBLIC_KEY0` public key, you can run the `Tools/cryptotools.py` Python program.
+For example, this command generates files called `new.pub` and `new.json` file in `Tools/test_keys`:
+
+```sh
+python3 Tools/cryptotools.py  --genkey Tools/test_keys/new
+```
+
+To generate a rsa2048 private and public key, you can use OpenSSL:
+
+```sh
+openssl genpkey -algorithm RSA -out private_key.pem -pkeyopt rsa_keygen_bits:2048
+```
+
+Then you can create a public key from this private key:
+
+```sh
+# Convert private_key.pem to a DER file
+openssl rsa -pubout -in private_key.pem -outform DER -out public_key.der
+# From the DER file, generate a public key in hex format, seperated by commas
+xxd -p public_key.der | tr -d '\n' | sed 's/\(..\)/0x\1, /g' > public_key.pub
+```
+
+To use these keys you would modify your `.px4board` file to point `CONFIG_PUBLIC_KEY0` to the path of `new.pub` that was generated via `cryptotools.py`, and set `CONFIG_PUBLIC_KEY1` to point to the file location of `rsa2048.pub`.
```

Comment:
What is the purpose of new.pub? We have an RSS key generated with open ssl, so what is the file for/what does it do?

Reply:
new.pub is the exchange key header file that is associated with the exchange key, while the RSS key generated with openssl is the encryption key.

Reply:
The exchange key is the key used to encrypt the symmetric key right? So new.pub is what encrypts the symmetric key?

What's the encryption key? That's a new term.

Reply:
The encryption key is used to encrypt the log itself. The CONFIG_PUBLIC_KEY param associated with it is determined by [SDLOG_KEY](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SDLOG_KEY).  However, I do see that [SDLOG_KEY](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SDLOG_KEY) says,

> Selects the key in keystore, used for encrypting the log. When using a symmetric encryption algorithm, the key is generated at logging start and kept stored in this index. For symmetric algorithms, the key is volatile and valid only for the duration of logging. The key is stored in encrypted format on the sdcard alongside the logfile, using an RSA2048 key defined by the SDLOG_EXCHANGE_KEY

Since XChaCha20 is a symmetric key algorithm, would this mean that it is useless to set an encryption key at the associated CONFIG_PUBLIC_KEY keystore, since encryption keys are volatile and change each time a new log is created?

Reply:
This is precisely my point. The only key we should to supply is the RSA key that encrypts the symmetric key (which is generated). 

I had assumed that this key was the key generated using open SSL (just because that sounds like the right way to do public key infrastructure than some random script we supply). But you're saying it is new.pub based on the default settings from encrypt. 

We should find out the intent - it might be that the openssl key is actually intended to be used as the exchange key, or that one of the keys has a completely different intended purpose and happens to be mapped by one of these two params. We really can't "guess".

Reply:
@jlaitine Can you shed light on this?

In summary, [`cryptotest.px4board`](https://github.com/PX4/PX4-Autopilot/blob/main/boards/px4/fmu-v5/cryptotest.px4board) sets the following keys which are stored in the key store.

```config
CONFIG_PUBLIC_KEY0="../../../Tools/test_keys/key0.pub"
CONFIG_PUBLIC_KEY1="../../../Tools/test_keys/rsa2048.pub"
```

Which of these is supposed to be used for the SDLOG exchange key (which encrypts the symmetric key), and what is the purpose of the other file?

The [SDLOG_EXCH_KEY ](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SDLOG_EXCH_KEY) param is value `1` which you would expect to map to `CONFIG_PUBLIC_KEY1` (0 indexed), so I think rsa2048.pub is the exchange key.

As per [this discussion ](https://github.com/PX4/PX4-user_guide/pull/3401#discussion_r1804127899) I think the other key is unrelated?

Reply:
Hi!

This is a good point. The "key0.pub" is indeed not related to the logfile encryption at all! This "key0.pub" actually originates from being the key used by the bootloader to verify the firmware signature. And obviously, when the public keys are compiled in to the code like it is done in the "stub_keystore", it doesn't make much sense in this context. I will still shed some light on why it was left in here..

- I some of the projects in which I am working, there is a real secure keystorage in the device (the stub_keystore is replaced by another hw driver which accesses keys stored in hw); that is, both the signature verification key and the RSA KEK are stored in the same place, thus they are accessed the same way and indexed always with different indexes no matter whether you are in bootloader or in PX4.
- When I started this work, I wanted to model this even when having the keys compiled in to the image - I just configured the "stub_keystore" the same way both in nuttx-based bootloader and in px4; so using the keys looked as if there was a real keystore with two keys
- Since key0.pub is an ed25519 elliptic curve signature verification key, it can be used for signature verification also inside PX4, not just in bootloader. The support for signature verification is actually also there built in to the sw_crypto driver, so it can be directly used in other application code, if needed.

I hope this clarifies, and I apologize the confusion which this caused. I understand that currently the "key0.pub" is not used for anything in upstrem PX4 - and it has nothing to do with logfile encryption.